Allow for heterogenous List and Dict values + Improve container typing algorithm

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57137**

Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[torch.Tensor]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

Commit:
Allow for heterogenous Dict values

[ghstack-poisoned]

gmagogsfm commented on test/jit/test_typing.py:
```diff
@@ -73,11 +73,17 @@ def test_dict_tensor_key(a, t):
+                d["bar"] = torch.add(t, t)
+            return d
+
+        self.checkScript(fn, (torch.rand(2, 3),))
```
Optional:

Use filecheck to verify type is correct in IR

ansley replied:
Good suggestion, added

gmagogsfm commented on test/jit/test_typing.py:
```diff
@@ -73,11 +73,17 @@ def test_dict_tensor_key(a, t):
-            @torch.jit.script
-            def foo():
-                new_item = {'score': [1.0], 'ys': [1, 2, 3]}
+    def test_dict_value_type_refinement_defaults_to_Any(self):
```
Are other ways of constructing heterogenous-value dicts supported too? like from dict comprehension, from gradually inserting different types of values etc.

ansley replied:
I added some new logic and test cases. Please take a look at the test cases I commented out and let me know whether you think we should support that too

ansley replied:
cc @gmagogsfm Reminder to look at those commented-out test cases :)

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3675,24 +3675,25 @@ struct to_ir {
+
+        for (size_t i = 0, N = values.size(); i < N; ++i) {
+          std::stringstream ss;
+          if (!values[i]->type()->isSubtypeOfExt(value_type, &ss) &&
+              !value_type->isSubtypeOfExt(values[i]->type(), &ss)) {
+            value_type = AnyType::get();
```
In case of `value_type` is a subtype of type of following values, should we update `value_type` to the super type?

Additionally, it looks like what you need is `unifyTypes` here, which would naturally return `Any` if types do not have subtyping relationship.

ansley replied:
Fixed, good call

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3675,24 +3675,25 @@ struct to_ir {
+        for (size_t i = 0, N = values.size(); i < N; ++i) {
+          std::stringstream ss;
+          if (!keys[i]->type()->isSubtypeOfExt(key_type, &ss) &&
+              !key_type->isSubtypeOfExt(keys[i]->type(), &ss)) {
```
Same here, should `key_type` be updated to super type?

ansley replied:
This might be a good follow-up project, but it's a lot bigger than it seems on the surface. The reason we restrict the allowable keys to a small set of types is that we generate a unique schema for each type

Commit:
Update on "Allow for heterogenous Dict values"


This PR corrects and expands our typing algorithm for unannotated, non-empty dicts. Previously, to verify type correctness for an unannotated, non-empty dict, we had gotten the type of the first key/value in the dict, then checked if each following key/value was a subtype of the first type. That's too restrictive--what if the first key was a subtype of the second key? Instead, we should type the dict by getting the greatest supertype of all the given keys or all the given values. 

We need slightly different rules for keys and values, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype.

This does not break backwards compatibility, because the default key/value type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty dicts are also still typed as `Dict[str, Tensor]`.

This PR unblocks two engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous Dict values"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the greatest supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1320,13 +1320,41 @@ struct to_ir {
+      bool use_old = out->type()->isSubtypeOf(lt->getElementType());
+
+      // If we have `use_old && use_new`, we just keep the old element
+      // type. Otherwise...
```
Is this comment incomplete?

ansley replied:
Thanks for pointing out that this comment is unclear. The "Otherwise..." meant "Otherwise, we do the logic directly below this comment". I'll change it!

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3533,6 +3593,47 @@ struct to_ir {
+  Value* emitListLiteral(ListLiteral ll, TypePtr type_hint) {
+    auto values = getValues(ll.inputs(), /*maybe_unpack=*/true);
+
+    // determine the element type of the list
+    // if we have a type hint of List[T], use T
+    // if the list is non-empty use type_of(list[0])
+    // otherwise assume it is List[Tensor]
```
This comment is no longer accurate

ansley replied:
Thanks!

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1320,13 +1320,41 @@ struct to_ir {
+
+      // Is the current element type the same type or a supertype of the
+      // old element type?
+      bool use_new = lt->getElementType()->isSubtypeOf(out->type());
+      // Is the old element type the same type or a supertype of the
+      // current element type?
+      bool use_old = out->type()->isSubtypeOf(lt->getElementType());
+
+      // If we have `use_old && use_new`, we just keep the old element
+      // type. Otherwise...
+      if (!use_old && use_new) {
+        list_value->setType(ListType::create(out->type()));
+      } else if (!use_old && !use_new) {
+        list_value->setType(ListType::create(AnyType::get()));
+      }
+
```
Good point

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3689,24 +3751,23 @@ struct to_ir {
-            }
+        for (size_t i = 0, N = values.size(); i < N; ++i) {
+          std::stringstream ss;
+          if (!unifyTypes(keys[i]->type(), key_type, /*default_to_any=*/false)) {
```
Good catch, thanks!

Commit:
Update on "Allow for heterogenous List and Dict values"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the greatest supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

gmagogsfm commented on test/HowToWriteTestsUsingFileCheck.md:
```diff
@@ -79,6 +79,9 @@ annotations from the example above one would write:
 
 * `CHECK: <pattern>`
   Scans the input until `PATTERN` is found. Fails if the pattern is not found.
+* `CHECK-NEXT: <pattern>`
+  Scans the input on the line immediately following the previous CHECK until
+  `PATTERN` is found. Fails if the pattern is not found on that line.
```
Consider making separate PR for changes that are not relevant to the functionality of PR

ansley replied:
Gotcha, will do that next time

gmagogsfm reviewed (changes requested):
There seems to be a few real test failures and lint warnings, could you push another version with them fixed?

Commit:
Update on "Allow for heterogenous List and Dict values"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the greatest supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

Commit:
Update on "Allow for heterogenous List and Dict values"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the greatest supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

Commit:
Update on "Allow for heterogenous List and Dict values"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the greatest supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1320,13 +1320,31 @@ struct to_ir {
+      // type
+      std::stringstream ss;
+      if (type_hint && !out->type()->isSubtypeOfExt(type_hint->expect<ListType>()->getElementType(), &ss)) {
+        throw ErrorReport(loc) << ss.str() << "List type annotation did"
```
`ss` contains the reason for not satisfying subtype relationship, it should ideally be at the end of error message.

ansley replied:
thanks!

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1320,13 +1320,31 @@ struct to_ir {
+
+      // If we didn't have a type annotation, let the type begin as the
+      // first element in the list
       if (!type_set) {
```
Why is this conditional type inference needed when the next step is to infer unified type unconditionally?

ansley replied:
If we didn't have a type annotation, the type of the list would be set to `Tensor`. We don't want to unify this default type with the actual elements in the list. I've updated the comment to explain this edge case

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3689,24 +3743,26 @@ struct to_ir {
+              values.begin(),
+              values.end(),
+              value_type,
+              [&](TypePtr a, Value* b) {
```
Good catch! Please take a look at my new algorithm

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the greatest supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the greatest supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3533,6 +3573,59 @@ struct to_ir {
+      }
+
+      if (*rhs_type == AnyType::get()) {
+        TORCH_WARN(
```
Done

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
+          // List of possible parent types
+          std::vector<TypePtr> seen;
+
+          auto get_greater_type =
+              [](const TypePtr t1, const TypePtr t2) -> c10::optional<TypePtr> {
+            if (t1->isSubtypeOf(t2)) {
+              return t2;
+            } else if (t2->isSubtypeOf(t1)) {
+              return t1;
+            } else {
+              return c10::nullopt;
+            }
+          };
```
This looks very similar to `unifyTypes(t1, t2, /*default_to_any=*false/)`, could you clarify the differences?

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3806,8 +3926,8 @@ struct to_ir {
 
-  // Desugars slice indexing: tensor[begin:end] -> tensor.slice(dim, begin, end,
-  // 1)
+  // Desugars slice indexing: tensor[begin:end] -> tensor.slice(dim, begin,
```
Changes that start from here to end of file seem to be lint only, could you move them to another PR?

ansley replied:
Thanks, I'll remove them

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
-        checkTypeOfValues(key_type, "keys", keys, key_trees);
-        checkTypeOfValues(value_type, "values", values, value_trees);
+
+          TypePtr inferred_type = seen.size() == 1 ? seen[0] : AnyType::get();
```
Could you clarify how this condition works for the motivating use case on line 3773?

It looks like that, at end of algorithm, `seen` would be [T_1, T_1], which would fail check of `seen.size() == 1`, making the final type `Any`. If this is indeed the case, we can change type of `seen` to `std::unordered_set`, which fixes this issue by automatically deduping elements. Additionally, `unordered_set` has faster look up time, which makes it easier to check if a `type` has been visited before.

ansley replied:
Good point. I'm actually going to open a PR to change `unifyTypeList` to this more forward compatible version (anticipating a future in which we have a less flat type hierarchy)

gmagogsfm replied:
Could you clarify how "changing unityTypeList" can address the incorrect result for motivating use case?

ansley replied:
This algorithm is basically doing what `unifyTypeList` is doing; it just accounts for the subtyping relationships you pointed out. Because the number of types in our types system is pretty small, changing `unifyTypeList` from the current `O(N)` algorithm to a forward-compatible `O(N^2)` algorithm would bolster us against the future without causing any real performance degradation. If we introduce polymorphic types in the next year or so and our type system expands greatly, we would need the `O(N^2)` algorithm anyway. So why not go ahead and change it?

We should also make our list emission forwards compatible (which it currently isn't). The new `unifyTypeList` could be used in all four sections of `ir_emitter.cpp` that this PR touches.

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
+
+        if (!values.empty()) {
+          // Get the smallest common supertype of all the values. We
+          // need to use the O(N^2) algorithm instead of using
+          // `std::accumulate` to account for the case that two types
```
```suggestion
          // need to use the O(N^2) algorithm to account for the case that two types
```

ansley replied:
fixed

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
+                      type_hint->expect<DictType>()->getValueType()),
+              "Type hint for dict values was ",
+              type_hint->repr_str(),
+              " but the greatest common supertype of the given values",
```
IIUC, this can only happen when there is at least one value that is not a sub-type of type hint. Is this accurate? If so, maybe we could make the error message simpler and more actionable if we loop through all value types and check their type relationship with type hint. That way, we would be able to tell user which exactly value offends typing rules, instead of erroring on an inferred supertype that user may not have specified themselves.

ansley replied:
fixed

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
-                  << type->repr_str() << " but found "
-                  << values[i]->type()->repr_str() << " instead.\n"
-                  << ss.str();
+        for (size_t i = 0, N = keys.size(); i < N; ++i) {
```
```suggestion
        for (size_t i = 0, n = keys.size(); i < n; ++i) {
```

ansley replied:
fixed

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1320,13 +1320,35 @@ struct to_ir {
+      ListTypePtr lt = list_value->type()->expect<ListType>();
+      auto unified = unifyTypes(
+          lt->getElementType(), out->type(), /*default_to_any=*/true);
+      list_value->setType(ListType::create(*unified));
```
Good point

ansley commented on test/jit/test_typing.py:
```diff
@@ -73,11 +74,80 @@ def test_dict_tensor_key(a, t):
+                    l[0] = (tup4[0], torch.add(t, t))
+            return l
+
+        self.checkScript(fn, (torch.rand(2, 3),))
```
I had no idea! Thanks!

gmagogsfm replied:
No worries.

On a side note, there is no need to hurry in replying to comments, take your time to fix implementation, wait for tests/lint to pass, push to Github and only then reply in comment about what was done to address the issue. Replying early doesn't really add much information. Also, please refrain from resolving comment thread, which can cause an issue to get lost in communication. Ideally only reviewer should resolve comment thread after reaching agreement on the fix.

ansley replied:
Thanks also for the side note! I'd been using the PR comments as a task list in the past, where a "resolve conversation" meant "done with the task". Moving forward, I'll use PR comments as you're suggesting! It makes a lot of sense!

You and I should talk offline about writing up a "Frontend GitHub Etiquette" doc for new hires. Some of the other reviewers I've worked with during my time at FB have different preferences (e.g. they only want me to silently hit "resolve conversation" unless I disagree with what they said). I think both ways are good, but we should make sure that everyone on the team is on the same page.

ansley commented on test/jit/test_typing.py:
```diff
@@ -73,11 +74,80 @@ def test_dict_tensor_key(a, t):
+
+        graph = torch.jit.script(fn).graph
+
+        FileCheck().check(r"(str, Any)[]").run(graph)
```
Perhaps this is proof that we should change '(' to `Tuple`... :)

The real type of the list is `List[Tuple[str, Any]]`. I'll change it to be the full `ListConstruct` line!

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1320,13 +1320,35 @@ struct to_ir {
+      if (type_hint &&
+          !out->type()->isSubtypeOfExt(
+              type_hint->expect<ListType>()->getElementType(), &ss)) {
+        throw ErrorReport(loc) << "List type annotation did not match "
```
Done! On a side note, I had to allow this behavior in `emitListLiteral` for BC purposes, but I think it really should be an error. How would you feel if I made it an error and simply changed our broken tests? I could stack another PR on top of this as a Better Engineering effort

gmagogsfm replied:
It does seem like a real error. Which test is broken?

ansley replied:
Yes, I deleted [this](https://github.com/pytorch/pytorch/pull/57137/files#diff-377406d8c2c4a9c34da88772f543dc94a0e81753a0fcb6208cdc579a057e627fL574) test case. It seems to me like what the code tests for is a mistake, but this may be me misunderstanding `torch.jit.interface` (there's no documentation for it at all). Please take a look before I ship this

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1365,10 +1387,33 @@ struct to_ir {
+      // annotatated key/value types
+      if (type_hint) {
+        DictTypePtr dict_type_hint = type_hint->expect<DictType>();
+        if (!k->type()->isSubtypeOf(dict_type_hint->getKeyType()) ||
+            !v->type()->isSubtypeOf(dict_type_hint->getValueType())) {
+          throw ErrorReport(loc) << "Dict type annotation did not match "
+                                 << "the types of the actual dict items";
```
All good points, thanks

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1365,10 +1387,33 @@ struct to_ir {
+          dt->getValueType(),
+          v->type(),
+          /*default_to_any=*/true);
+      dict_value->setType(DictType::create(k->type(), *unified));
```
Done

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[str]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1366,10 +1415,95 @@ struct to_ir {
+              << ss.str();
+        }
+
+        if (!is_key_subtype && !is_value_subtype) {
+          throw ErrorReport(dc)
+              << "Dict type annotation `" << dict_type_hint->repr_str()
+              << "` did not match the types of the "
+              << "actual dict items";
+        }
```
This check is superceded by the check of 1457-1459 I think.

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1366,10 +1415,95 @@ struct to_ir {
+      // to allow for the case that `*unified` is a subtype of
+      // the value type given by `type_hint`
+      if (!type_hint) {
+        dict_value->setType(DictType::create(k->type(), *unified));
```
For key type inference, is it:

1) check if there is type annotation, if so, check if all key types are subtypes of annotation. Then use annotation
2) if there is no type annotation, use key type of first pair in dictionary. 

Is this correct?

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[torch.Tensor]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[torch.Tensor]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[torch.Tensor]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[torch.Tensor]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[torch.Tensor]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[torch.Tensor]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[torch.Tensor]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).

Commit:
Update on "Allow for heterogenous List and Dict values + Improve container typing algorithm"


Differential Revision: [D28231839](https://our.internmc.facebook.com/intern/diff/D28231839)

This PR corrects and expands our typing algorithm for unannotated, non-empty dicts and lists. Previously, to verify type correctness for an unannotated, non-empty container, we had gotten the type of the first element in the container, then checked if each following element was a subtype of the first type. That's too restrictive--what if the first element were a subtype of the second element? Instead, we should type the container by getting the smallest common supertype of all the given elements.

We need slightly different rules for keys and values in dicts, though: because the set of key types is restricted, finding two key types that cannot be unified should cause an error. On the other hand, the set of value types is not restricted, so we should be able to use `Any` as a valid supertype. We need to keep the set of keys restricted since the keys are used to generate and match schemas.

This does not break backwards compatibility, because the default element type is the smallest supertype of all the given types. So, if someone creates an unannotated dict where the keys are all `str` and the values are all `torch.Tensor`, the dict will be inferred to `Dict[str, Tensor]` just like it was before. Empty lists are still typed as `List[torch.Tensor]`, and empty dicts are still typed as `Dict[str, Tensor]`.

This PR unblocks three engineers on an FB-internal team and improves FX-TorchScript compatibility.

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28231839).
